mypy: enable option `strict_optional` for 'pylib' (#3317)

davidculley opened the pull request:
Fixed all errors within the directory `pylib` by simply adding assertions. These can be replaced later on by better solutions. But we can now set the option `strict_optional` to `true` for the entire directory `pylib`. 

Why this option should be `true`: https://mypy.readthedocs.io/en/stable/config_file.html#confval-strict_optional

dae commented:
```diff
diff --git a/pylib/anki/models.py b/pylib/anki/models.py
index 0d7024933..d3dd5c886 100644
--- a/pylib/anki/models.py
+++ b/pylib/anki/models.py
@@ -134,9 +134,7 @@ class ModelManager(DeprecatedNamesMixin):
 
     def current(self, for_deck: bool = True) -> NotetypeDict:
         "Get current model. In new code, prefer col.defaults_for_adding()"
-        model_id = self.col.decks.current().get("mid")
-        assert model_id is not None
-        notetype = self.get(model_id)
+        notetype = self.get(self.col.decks.current().get("mid"))
         if not for_deck or not notetype:
             notetype = self.get(self.col.conf["curModel"])
         if notetype:
@@ -154,7 +152,7 @@ class ModelManager(DeprecatedNamesMixin):
         except NotFoundError:
             return None
 
-    def get(self, id: NotetypeId) -> NotetypeDict | None:
+    def get(self, id: NotetypeId | None) -> NotetypeDict | None:
         "Get model with ID, or None."
         # deal with various legacy input types
         if id is None:
diff --git a/pylib/anki/tags.py b/pylib/anki/tags.py
index af1f2c6f9..ce6335991 100644
--- a/pylib/anki/tags.py
+++ b/pylib/anki/tags.py
@@ -121,10 +121,9 @@ class TagManager(DeprecatedNamesMixin):
     def rem_from_str(self, deltags: str, tags: str) -> str:
         "Delete tags if they exist."
 
-        def wildcard(pat: str, repl: str) -> Match:
+        def wildcard(pat: str, repl: str) -> Match | None:
             pat = re.escape(pat).replace("\\*", ".*")
             return_value = re.match(f"^{pat}$", repl, re.IGNORECASE)
-            assert return_value is not None
             return return_value
 
         current_tags = self.split(tags)
```

davidculley commented:
@dae Thanks a lot for the super helpful fix. 🙂

dae commented:
Making the code generator insert them would be tricky. I suggest

```diff
diff --git a/.mypy.ini b/.mypy.ini
index 96dae8669..f15a768c2 100644
--- a/.mypy.ini
+++ b/.mypy.ini
@@ -26,6 +26,8 @@ disallow_untyped_defs = True
 strict_optional = true
 [mypy-aqt.*]
 strict_optional = false
+[mypy-_aqt.*]
+strict_optional = false
 [mypy-anki.importing.*]
 disallow_untyped_defs = False
 [mypy-anki.exporting]
```

dae commented:
Please note that Anki currently fails to start, but you've added an assert in a location where the value can be none. We'll need to do a quick pass over each of the asserts that have been added to double-check the values should never be None, as there may be other instances where the types were not correctly indicating that the value was nullable.

davidculley commented:
> I suggest

Good idea. Thanks for helping me out. 👍

> We'll need to do a quick pass over each of the asserts that have been added to double-check the values should never be None

Absolutely. I have no idea if any of the assertions I added are actually correct. I don't understand the codebase well enough to know that. My plan was to first make all mypy errors disappear, then to make the pipeline succeed if necessary, and then go over all assertions I added and see if they're actually correct. If any assertion is wrong, replace it with a real solution and repeat the step of making the pipeline succeed. Kind of like an "assertion-driven development". 😅

So with your help, I got the pipeline to succeed now. 🙂 Now we can look which assertions need to be replaced.

davidculley commented:
To make the code changes easier to review for you, I moved some changes into their separate pull request (#3364).

The assertions become easier to review if you first merge #3364. 🙂

dae commented:
I'd appreciate it if you could make an attempt to review these yourself first - Anki wouldn't even start until the latest commit I pushed, and when I started reviewing the changes, I quickly encountered changes that don't appear to be correct, such as

```
    def get_image_for_occlusion(self, path: str | None) -> GetImageForOcclusionResponse:
        assert path is not None
```

dae commented:
Do you plan to return to this? I can keep it around if you're currently busy, as it would be a nice change.
